### DISCUSSION TITLE
Error handling when using SSH and experiencing a timeout?

### DISCUSSION THREAD

**Discussion by SelfhostedPro** [apocas/dockerode#731] [Q&A]
Hey, I'm working on a new version of my docker ui but I'm running into issues when trying to iterate through multiple servers. I've got a dictionary of `Docker` objects (ie. created using `new Docker()`). When the server is on/accessible it works as expected. When the server is off, ssh2 throws an error and I'm unable to easily catch it. I've got the following function I'm using:
```ts
  async getContainers(): Promise<ServerContainers> {
    const servers: ServerDict =
      await this.serversService.getServersFromConfig();
    const serverKeys = Object.keys(servers);
    // Get containers from all servers in config
    const serverPromises: Promise<Container[]>[] = serverKeys.map((name) =>
      servers[name].listContainers({ all: true }).then(normalizeContainers),
    );
    // Wait for containers to resolve
    const containerArrays = await Promise.all(serverPromises);
    // Assign each container array to it's server
    return serverKeys.reduce((acc, serverName, index) => {
      acc[serverName] = containerArrays[index];
      return acc;
    }, {} as { [serverName: string]: Container[] });
  }
  ```
  
  Even inside of a try/catch block I'm getting the following error:
  ```
  node_modules/.pnpm/ssh2@1.12.0/node_modules/ssh2/lib/client.js:1116
          const err = new Error('Timed out while waiting for handshake');
                      ^
Error: Timed out while waiting for handshake
    at Timeout._onTimeout (/Users/stephen/dev/Personal/yacht-api/node_modules/.pnpm/ssh2@1.12.0/node_modules/ssh2/lib/client.js:1116:23)
    at listOnTimeout (node:internal/timers:569:17)
    at processTimers (node:internal/timers:512:7) {
  level: 'client-timeout'
}
```

Any ideas on how I could catch this or does docker-modem need to utilize the `.on('error'` side of things? (ie. https://github.com/SelfhostedPro/yacht-api/blob/9412777b4e9d929dcfdaf4f795b5abb3d5e8cdf8/server/src/util/sshKeyManager.ts#L106-L125)

Link to code/repo: https://github.com/SelfhostedPro/yacht-api/blob/9a56304819149e00d83e9af7b5eec33602deb553/server/src/containers/containers.service.ts#L33-L48

**Comment by apocas** [comment]
Yes I believe docker-modem needs to control this error as well.

https://github.com/apocas/docker-modem/commit/f479f7b3f41f13bc274dfdb56e5b99c2ab0746c3#diff-701f98c1437169db3aa700d64c6e401c99f50589523af7ad23a4033469850888R25

Could you please test this branch? Check the small change I did.

**Reply by d0Nt** [reply]
I had similar issue and this change works
